Code/ForceField/ANI/AtomicContrib.cpp

```diff
@@ -0,0 +1,469 @@
+const double A_zeta = 32.0;
+const int len_R_Rs = 16;
+const double R_Rs[len_R_Rs] = {
+    0.9,  1.16875, 1.4375, 1.70625, 1.975, 2.24375, 2.5125, 2.78125,
+    3.05, 3.31875, 3.5875, 3.85625, 4.125, 4.39375, 4.6625, 4.93125};
+const int len_A_thetas = 8;
+const double A_thetas[len_A_thetas] = {0.19634954, 0.58904862, 0.9817477,
+                                       1.3744468,  1.7671459,  2.1598449,
+                                       2.552544,   2.9452431};
+const int len_A_Rs = 4;
+const double A_Rs[len_A_Rs] = {0.9, 1.55, 2.2, 2.85};
+const double radial_prefactor = 0.25;
+const double inner_product_prefactor = 0.95;
+
+ANIAtomContrib::ANIAtomContrib(ForceField *owner, VectorXi &speciesVec,
+                               std::string model) {
+  PRECONDITION(owner, "This ForceField contrib needs an owner")
+  dp_forceField = owner;
+  // load element mapping and atomic self-energies
+  this->paramDir = model;
+  std::ifstream selfEnergyFile(paramDir + "/selfEnergies");
+  if (!selfEnergyFile.is_open()) {  // try default location
+    std::string rdbase = getenv("RDBASE");
+    paramDir = rdbase + "/Code/ForceField/ANI/Params/" + model;
+    selfEnergyFile.open(paramDir + "/selfEnergies");
+  }
+  PRECONDITION(selfEnergyFile.is_open(),
+               "Missing ANI energy file " + paramDir + "/selfEnergies")
+
+  std::map<int, std::string> elements_by_index;
+  std::map<std::string, int> indices_by_element;
+  // parse atomic energy lines of the form "H,0=-0.60095298000"
+  std::string next;
+  while (!selfEnergyFile.eof()) {
+    std::getline(selfEnergyFile, next, ',');
+    std::string element = next;
+    std::getline(selfEnergyFile, next, '=');
+    int index = atoi(next.c_str());
+    std::getline(selfEnergyFile, next);
+    double E = atof(next.c_str());
+    elements_by_index[index] = element;
+    indices_by_element[element] = index;
+    this->d_selfEnergies.push_back(E);
+  }
+  selfEnergyFile.close();
+  n_elements = elements_by_index.size();
+  n_features =
+      n_elements * ((n_elements - 1) / 2.0 + 1) * len_A_Rs * len_A_thetas +
+      n_elements * len_R_Rs;
+
+  // make d_atomTypes using indices_by_element and speciesVec
+  auto table = RDKit::PeriodicTable::getTable();
+  this->d_atomTypes.resize(speciesVec.size());
+  for (long i = 0; i < speciesVec.size(); i++) {
+    std::string element = table->getElementSymbol(speciesVec[i]);
+    if (indices_by_element.find(element) == indices_by_element.end()) {
+      throw ValueErrorException("Element not Supported: " + element);
+      return;
+    }
+    this->d_atomTypes[i] = indices_by_element[element];
+  }
+
+  // load parameters by model number in ensemble, by element, and by layer
+  int model_i = 0;
+  while (true) {
+    // get ensembleSize by counting files in param dir
+    std::string paramFile =
+        paramDir + "/model" + std::to_string(model_i) + ".bin";
+    FILE *fp = fopen(paramFile.c_str(), "rb");  // check that file exists
+    fclose(fp);
```

Comment:
You could be calling fclose on a null ptr, move this after the break.

Reply:
Good point. I was worried about leaving it open, but of course if it's null then it's not open by definition.

Reply:
Moving fclose() seems to have fixed the Ubuntu test failures. Nice catch!

If I'm reading the Azure output correctly, the Mac test has now failed in the build step, and not at a file I recognize.

```
[ 91%] Linking CXX executable moldraw2DRxnTest1
clang-10: error: unable to execute command: Segmentation fault: 11
clang-10: error: linker command failed due to signal (use -v to see invocation)
make[2]: *** [Code/GraphMol/MolDraw2D/moldraw2DRxnTest1] Error 254
make[1]: *** [Code/GraphMol/MolDraw2D/CMakeFiles/moldraw2DRxnTest1.dir/all] Error 2
```

My Mac build process prints this perhaps-related but non-fatal message:

```
/Library/Developer/CommandLineTools/usr/bin/ranlib: file: /Users/stevenso/Downloads/rdkit/manangoel/rdkit/lib/libRDKitMolDraw2D_static.a(MolDraw2DJS.cpp.o) has no symbols
```

In case it matters, my cmake command is:
```
cmake -DPYTHON_INCLUDE_DIR=$PYROOT/include/python3.8/  \
  -DRDK_BUILD_AVALON_SUPPORT=OFF \
  -DRDK_BUILD_CAIRO_SUPPORT=ON \
  -DRDK_BUILD_INCHI_SUPPORT=ON \
  -DBoost_NO_BOOST_CMAKE=ON ..
```

Reply:
Whoa, that's a clang linker segment fault.  I think that's a random error...

Reply:
Is there a way to retry the build without pushing an empty commit?

Reply:
@bp-kelley anything we can do to test if it is a random error?

Reply:
There might be a way to restart a build, but since we squash pushing a commit isn’t too bad.